### PR TITLE
[GH-912]:Fixed issue "Handle error case where issuetypes are not available" of Jira plugin 

### PR DESCRIPTION
**Pull request by Kshitij-Katiyar**
**Summary** 

- In some cases, while getting the metadata for projects issueTypes were inaccessible due to some permission schemes. So now changed the pluginAPI ```/api/v2/get-jira-project-metadata``` to get the project list first and then get the issueTypes for each project to not show the projects whose issueTypes are inaccessible.

- Reference https://community.mattermost.com/_redirect/pl/zgocguq9wtgbtmztrctztrkrhh

**Issue** 

- [Issue #912](https://github.com/mattermost/mattermost-plugin-jira/issues/912)

### CODE REVIEW

File: server/issue.go [hunk 1]
```diff
@@ -521,12 +521,18 @@ func (p *Plugin) httpGetJiraProjectMetadata(w http.ResponseWriter, r *http.Reque
 	projects := []utils.ReactSelectOption{}
 	issues := map[string][]utils.ReactSelectOption{}
 	for _, prj := range plist {
+		issueTypeList, err := p.GetIssueTypes(types.ID(instanceID), types.ID(mattermostUserID), prj.ID)
+		if err != nil {
+			p.API.LogError("Failed to get issue types for project.", "ProjectKey", prj.Key, "Error", err.Error())
```
Comment (mickmister) [inline server/issue.go:526]: @Kshitij-Katiyar This list of projects is potentially very large. With the current implementation here, we will run into an `n+1` problem.

What do you think about my suggestion here? https://github.com/mattermost/mattermost-plugin-jira/issues/912#issuecomment-1430341080

> If we run into the error (does it tell us which projects?), then fetch the projects without expanding the issue type. Then individually fetch issue types for each project. This is not performant, but may be necessary if we don't know which projects have issue types.

I'm thinking we should first attempt to fetch all projects with one request like we used to be doing, and if that throws an error then fetch them without the expand, and then loop through them and fetch issue types like this, but only when that error occurs.

Another question: Are we able to fetch the projects without issue types, and determine if they have issue types somehow, without triggering the Jira error?

Comment (mickmister) [reply server/issue.go:526]: Please take into account how this affects Jira Cloud. We don't want to decrease performance there while fixing a bug for Jira Server, so we should by default still be fetching everything with one request for Jira Cloud.

**Review by mickmister — CHANGES_REQUESTED** [review]
Thanks for implementing this @Kshitij-Katiyar :+1: I have a few concerns related to performance

File: server/issue.go (hunk 1 above)
Comment (Kshitij-Katiyar) [reply server/issue.go:526]: @mickmister No it does not tell us which project is triggering the error if we fetch the project list with issueTypes expanded.
Yes, we can first fetch the project with issueTypes expanded and if we run into an error then we can fetch issueTypes individually for each project. Just have one question here should we fetch in all errors or just when we will get this particular error?
And we cannot determine in the project list whether they have issueTypes or not without calling any additional API.

Comment (mickmister) [reply server/issue.go:526]: > Just have one question here should we fetch in all errors or just when we will get this particular error?

@Kshitij-Katiyar I trust your judgment on this. I'm leaning towards "just that one kind of error". Note that this error could be in another language than English, so hopefully there is an error id or something like that available.

**Comment by mickmister** [comment]
@Kshitij-Katiyar Are you able to share how to specifically configure a Jira server to reproduce this issue? Or possibly share the actual Jira server you're using to assert the code works?

cc @DHaussermann

**Comment by Kshitij-Katiyar** [comment]
> @Kshitij-Katiyar Are you able to share how to specifically configure a Jira server to reproduce this issue? Or possibly share the actual Jira server you're using to assert the code works?
> 
> cc @DHaussermann

@mickmister No I am not able to reproduce this particular with my Jira server (tried many permissions). For testing purposes, Just for the sake of Jira returning an error, I have created a wrong custom issue type for a project but the error it returns is not the same as this issue.

File: server/issue.go (hunk 1 above)
Comment (Kshitij-Katiyar) [reply server/issue.go:526]: @mickmister I am not able to find any error id associated with this but we can do this for this particular status code. 
One more question should I remove the changes for the cloud part?

Comment (mickmister) [reply server/issue.go:526]: @Kshitij-Katiyar It's probably best to take the defensive approach and include it for Cloud as well. One thing I want to emphasize is we should avoid the n+1 situation in any way possible. I realize it's required in some scenarios though.

Comment (mickmister) [reply server/issue.go:526]: @Kshitij-Katiyar I think we should continue fetching the issue types with `expand` for both Server and Cloud, and fallback on `GetIssueTypes` only when we need to

Comment (Kshitij-Katiyar) [reply server/issue.go:526]: sure @mickmister

File: server/issue.go [hunk 2]
```diff
@@ -502,10 +502,31 @@ func (p *Plugin) httpGetJiraProjectMetadata(w http.ResponseWriter, r *http.Reque
 
 	instanceID := r.FormValue("instance_id")
 
-	plist, connection, err := p.ListJiraProjects(types.ID(instanceID), types.ID(mattermostUserID))
+	plist, connection, err := p.ListJiraProjects(types.ID(instanceID), types.ID(mattermostUserID), true)
 	if err != nil {
-		return respondErr(w, http.StatusInternalServerError,
-			errors.WithMessage(err, "failed to GetProjectMetadata"))
+		// Getting the issue Types separately only when the status code returned is 400
+		if strings.Contains(err.Error(), "400") {
```
Comment (mickmister) [inline server/issue.go:508]: To limit the indentation below, we can return early here

```suggestion
		if !strings.Contains(err.Error(), "400") {
			return respondErr(w, http.StatusInternalServerError,
				errors.WithMessage(err, "failed to GetProjectMetadata"))
		}
```

**Review by mickmister — CHANGES_REQUESTED** [review]
THanks for addressing the concerns @Kshitij-Katiyar 

One remaining request to avoid using `LogError` for these errors

File: server/issue.go [hunk 3]
```diff
@@ -502,10 +502,31 @@ func (p *Plugin) httpGetJiraProjectMetadata(w http.ResponseWriter, r *http.Reque
...
+			if err != nil {
+				return respondErr(w, http.StatusInternalServerError,
+					errors.WithMessage(err, "failed to get the list of Jira Projects"))
+			}
+
+			var projectList jira.ProjectList
+			for _, prj := range plist {
+				issueTypeList, iErr := p.GetIssueTypes(types.ID(instanceID), types.ID(mattermostUserID), prj.ID)
+				if iErr != nil {
+					p.API.LogError("Failed to get issue types for project.", "ProjectKey", prj.Key, "Error", iErr.Error())
```
[the remark this answers is not in the document]
Comment (hanzei) [reply server/issue.go:519]: :+1: for debug

**Review by hanzei — APPROVED** [review]

**Review by mickmister — APPROVED** [review]
Awesome thanks @Kshitij-Katiyar :+1:

**Comment by DHaussermann** [comment]
Hi @Kshitij-Katiyar 
I've done a round of testing on this and there is still an issue with loading projects in the subscriptions and craete modal. 
When the list of projects contains a record where the test user does not have create permissions, even projects where both browse and create are granted are not usable for creating issues and subscriptions. 
Ex: 
Open the subscription modal when    
Project A --> Browse and Create access
Project B --> Browse access

I see an error when either project is selected. 
`{"timestamp":"2023-02-24 18:05:28.928 Z","level":"error","msg":"ERROR: ","caller":"app/plugin_api.go:976","plugin_id":"jira","Status":"500","Error":"Request failed. Please analyze the request body for more details. Status code: 400","Path":"/api/v2/get-create-issue-metadata-for-project","Method":"GET","query":"instance_id=http%3A%2F%2Fdkh-token.ngrok.io&project-keys=PEP"}`
After some investigation by @mickmister it was determined the solution here is only partially implemented https://github.com/mattermost/mattermost-plugin-jira/issues/912#issuecomment-1430341080. To narrow down the issue, Michael made a change on a separate branch for debugging purposes https://github.com/mattermost/mattermost-plugin-jira/blob/c40056257d013d2668fea2f285e590f04ac9bc99/server/client_server.go#L67  
With this change deployed Project A is usable but Project B shows no issue types available. In this case - Project B should not be shown in the drop down at all.  

Based on this, can you please implement a solution? It would also be useful to add more context on the error message received above as it seems to be received from Jira with no other context in the log.

**Comment by Kshitij-Katiyar** [comment]
@mickmister @hanzei @DHaussermann  I explored a bit on the issue keeping the above post of @DHaussermann in mind.
I tested the code for the jira server above and below 9.0.0 and on the cloud as well. Everything is working fine for the jira cloud.

There are mainly 2 problems I am facing to resolve this issue.
1) The first is to remove the projects from the dropdown itself where we don't have access to ```create issue```. When we open the subscription modal, an API ```get-jira-project-metadata```eventually which eventually calls the jira API ``` /rest/api/2/project``` to get the list of projects for an instance. So I thought we can expand the ```issueTypes``` parameter and see if we have access and if we don't in that case we can individually fetch the issueTypes for each projects and remove the projects we don't have access to.
But this API doesn't expand ```issueTypes``` for the jira server and I am attaching a screenshot for the supported values of expand parameter. The supported values are ```description,lead,url,projectKeys```.
Screenshot from 2023-03-01 19-40-16
Without this, the only solution I have is to iterate on every project every time if we don't want these projects in the dropdown itself. This could slow down the process a lot.

2) The second problem is for the Jira servers above 9.0.0 after we select a project even if we have access to this project if there is even 1 project in that list we don't have access to we are going to get the same error due to the code. 
https://github.com/mattermost/mattermost-plugin-jira/blob/e277ede06d0270c1ac3d3c3e116946c31df6db2e/server/client_server.go#L71
As we can see this code is just breaking and returning from the loop rather than just continuing with the next project in that list.
For now, I think we can fix this part and after the fix, users can see all the projects in the list and can select and use the projects they have access to but if they select the project they don't have access to we can just show them the error.

Please give your views on this.

**Comment by mickmister** [comment]
@Kshitij-Katiyar Thanks so much for digging into the problem

> but if they select the project they don't have access to we can just show them the error.

This is fine for now :+1:

cc @DHaussermann For the purposes of this PR, we _will_ show the projects the user does not have create permissions for, but will show an error to the user saying why they can't create a subscription for that project if they select that project. This defect is a product of the changes needed for the issue this PR is fixing.

As an improvement after this PR, maybe this API endpoint could help us figure out if a user has access to create issues before we populate the select box?

https://developer.atlassian.com/cloud/jira/platform/rest/v3/api-group-permissions/#api-rest-api-3-permissions-check-post

-----

@Kshitij-Katiyar It's hard to tell by your comment, are you blocked on anything or is the path forward here clear?

**Comment by Kshitij-Katiyar** [comment]
@mickmister Sure I was just asking whether I should implement the solution I suggested before doing the changes.  For now, I fix problem 2 in my above post of just showing the error and after implementing that I will explore more on the API you have suggested. Is it fine?

**Comment by mickmister** [comment]
@Kshitij-Katiyar That sounds good to me :+1: Let's get this merged 🙂

**Comment by DHaussermann** [comment]
> For the purposes of this PR, we will show the projects the user does not have create permissions for, but will show an error to the user saying why they can't create a subscription for that project if they select that project. 

Understood. 👍 

@mickmister and @Kshitij-Katiyar I'll tst this as soon as possible. 

@Kshitij-Katiyar please let me know once you have a new commit up to implement this. As of right now, I don't see anything new.

**Comment by mickmister** [comment]
@Kshitij-Katiyar I found another issue with the function we're discussing. Please see this commit for a related fix that should go in this PR: https://github.com/mattermost/mattermost-plugin-jira/commit/eb760151c999a8db03e8256f388041fafa505a5d. Note that the linked commit is on a debug branch, and it won't be going to master outside of your PR.

Essentially, we were fetching all projects when the client is requesting `CreateMetaInfo` for a specific project. The linked commit makes it so we only fetch the specified project. Also notice that some methods are renamed for more accurate representation of the return types. The `if err != nil { break }` issues still need to be addressed as you pointed out.

**Comment by Kshitij-Katiyar** [comment]
> @Kshitij-Katiyar I found another issue with the function we're discussing. Please see this commit for a related fix that should go in this PR: [eb76015](https://github.com/mattermost/mattermost-plugin-jira/commit/eb760151c999a8db03e8256f388041fafa505a5d). Note that the linked commit is on a debug branch, and it won't be going to master outside of your PR.
> 
> Essentially, we were fetching all projects when the client is requesting `CreateMetaInfo` for a specific project. The linked commit makes it so we only fetch the specified project. Also notice that some methods are renamed for more accurate representation of the return types. The `if err != nil { break }` issues still need to be addressed as you pointed out.

@mickmister I saw this too and was thinking of fixing it. Thanks for fixing it. Push the changes for the other issue as well.

**Comment by m1lt0n** [comment]
@mickmister @hanzei I've added a rereview request for the latest commits that fix 915

File: server/client_server.go [hunk 4]
```diff
@@ -60,67 +61,63 @@ func (client jiraServerClient) GetIssueInfo(projectID string) (*ProjectIssueInfo
...
+		return nil, resp, errors.Wrap(err, "failed to get create meta info")
 	}
 
-	for _, proj := range *projectList {
-		meta.Expand = proj.Expand
-		issueInfo, resp, err = client.GetIssueInfo(proj.ID)
+	for _, issueType := range issueInfo.Values {
+		apiEndpoint := fmt.Sprintf("%s%s/issuetypes/%s", APIEndpointCreateIssueMeta, proj.ID, issueType.Id)
+		req, rErr := client.Jira.NewRequest(http.MethodGet, apiEndpoint, nil)
 		if err != nil {
```
Comment (mickmister) [inline server/client_server.go:79]:
```suggestion
		req, rErr := client.Jira.NewRequest(http.MethodGet, apiEndpoint, nil)
		if rErr != nil {
```

**Review by mickmister — APPROVED** [review]
Thanks for addressing the issues @Kshitij-Katiyar :+1:

LGTM other than one typo I see here. @DHaussermann This is ready for review

**Comment by DHaussermann** [comment]
Tested and this is now wp

**Review by mickmister — APPROVED** [review]

**Review by DHaussermann — APPROVED** [review]
Tested and passed

All issues with this now see resolved. When the subscribe modal is opened.... 
- Project A has Browse and Create access for test user --> Works normally
- Project B has Browse access for test user and no create --> Error message now shows only when project is selected
- Project C has no browse or create access for test user --> project is no listed and no error about it is shown

I have also tested this issue create modal

LGTM!

Huge thanks @Kshitij-Katiyar for the effort on this.

**Review by hanzei — APPROVED** [review]

**Comment by hanzei** [comment]
@Kshitij-Katiyar Could you, please open a separate PR that cherry-picks these changes into  `release-3.2`? Thanks!

**Comment by Kshitij-Katiyar** [comment]
> @Kshitij-Katiyar Could you, please open a separate PR that cherry-picks these changes into `release-3.2`? Thanks!

@hanzei Cherry picked the commit https://github.com/mattermost/mattermost-plugin-jira/commit/a7f0e417ba67ba5a0113815cad2d6f655aceffc3 into the branch release-3.2 .There were some merge conflicts, but i have resolved them. Let me know if everything looks fine.

**Pull request merged**